internal/provider/data_ko_resolve.go

```diff
@@ -0,0 +1,150 @@
+	return &schema.Resource{
+		Description: "",
+
+		ReadContext: dataDryResolveConfigRead,
+
+		Schema: map[string]*schema.Schema{
+			FilenamesKey: {
+				Description: "Filenames, directorys, or URLs to files to use to create the resource",
+				Required:    true,
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				ForceNew:    true,
+			},
+			RecursiveKey: {
+				Description: "Process the directory used in -f, --filename recursively. Useful when you want to manage related manifests organized within the same directory.",
+				Optional:    true,
+				Type:        schema.TypeBool,
+				ForceNew:    true,
+			},
+			SelectorKey: {
+				Description: "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)",
+				Optional:    true,
+				Type:        schema.TypeString,
+				ForceNew:    true,
+			},
+
+			// Computed fields
+			ManifestsKey: {
+				Description: "A list of resolved manifests in a 'yamldecode'able format. Note that whitespaces and nil docs will be stripped from these results.",
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+			},
+		},
+	}
+}
+
+// dryResolver is an implementation of Resolver that doesn't actually build anything
+type dryResolver struct {
+	fo *options.FilenameOptions
+	so *options.SelectorOptions
+}
+
+func NewDryResolver(d *schema.ResourceData, meta interface{}) (*dryResolver, error) {
+	r := &dryResolver{
+		fo: &options.FilenameOptions{},
+		so: &options.SelectorOptions{},
+	}
+
+	if p, ok := d.Get(FilenamesKey).([]interface{}); ok {
+		r.fo.Filenames = StringSlice(p)
+	}
+
+	if p, ok := d.Get(RecursiveKey).(bool); ok {
+		r.fo.Recursive = p
+	}
+
+	if p, ok := d.Get(SelectorKey).(string); ok {
+		r.so.Selector = p
+	}
+
+	return r, nil
+}
+
+func (r *dryResolver) Resolve(ctx context.Context) (*Resolved, error) {
+	var resolveBuf bytes.Buffer
+	w := &nopWriteCloser{Writer: bufio.NewWriter(&resolveBuf)}
+
+	emptyBuilder, err := build.NewCaching(&emptyBuilder{})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := commands.ResolveFilesToWriter(ctx, emptyBuilder, &emptyPublisher{}, r.fo, r.so, w); err != nil {
+		return nil, err
+	}
+
+	if err := w.Flush(); err != nil {
+		return nil, err
+	}
+
+	return NewResolved(resolveBuf.Bytes())
+}
+
+func dataDryResolveConfigRead(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	r, err := NewDryResolver(d, meta)
+	if err != nil {
+		return diag.Errorf("building dry resolver: %v", err)
+	}
+
+	resolved, err := r.Resolve(ctx)
+	if err != nil {
+		return diag.Errorf("resolving: %v", err)
+	}
+
+	d.SetId(resolved.ID())
+	d.Set("manifests", resolved.Manifests)
+	return nil
+}
+
+var _ = build.Interface(&emptyBuilder{})
+
+type emptyBuilder struct{}
+
+// Build implements build.Interface
+func (*emptyBuilder) Build(ctx context.Context, ip string) (build.Result, error) {
+	return empty.Image, nil
```

Comment:
Does it help if this actually builds but doesn't publish? This means the output of this can include the real resolved digests, same as you'd get from `ko_resolve`, so when you actually publish later it's mostly a no-op.

Or come to think of it, can the data source also publish, and just not tag? If this works, maybe ko_resolve should just be a data source. Maybe ko_image should be too.

This also makes me think we need to be more careful about tagging in general in here, since we expect folks to `terraform plan` without deleterious side effects like updating tags. Pushing manifests by digest seems safe though.

Reply:
I actually really like that idea... and think it jives well with terraform verbiage. since `data.` loosely translates to "lookups", `data.ko_resolve` could be interpreted as "lookup the resolved manifests from disk but don't publish", and `resource.ko_resolve` could be more focused on publishing, and include attributes that deal with that (tags)